src/filesystem.cpp

```diff
@@ -761,33 +789,32 @@ void set_user_data_dir(std::string newprefdir)
 					{1, 17, 0},
 					_("Use absolute paths. Relative paths are deprecated because they are interpreted relative to $HOME"));
 			}
-			user_data_dir = home / newprefdir;
+			dir = home / temp;
 		}
 	}
-#else
+
+	return dir;
+}
+
+void set_user_data_dir(std::string newprefdir)
+{
+	[[maybe_unused]] bool relative_ok = false;
+
+#ifdef PREFERENCES_DIR
 	if(newprefdir.empty()) {
-		newprefdir = backupprefdir;
+		newprefdir = PREFERENCES_DIR;
```

Comment:
Basically, instead of setting `relative_ok = true` here, we could add something equivalent to the following:

```c++
if(newprefdir[0] == '~') {
    newprefdir = getenv("HOME") + newprefdir.substr(1);
}
```

That's not quite right, as the environment variable has a different name on Windows, but you get the idea, right? Maybe Boost even provides the function for us…

```c++
// Hypothetical function that might not exist or may have a different name
newprefdir = bfs::expand_user(PREFERENCES_DIR);
```

Reply:
Implemented it, though not in this way.

Reply:
If I understand correctly, that means it doesn't work on Windows anymore?

Though that's probably fine, as Windows is the least likely to care…

Reply:
Windows didn't make use of the `relative_ok` variable anyway, so it's just still not there. Windows instead has the special handling of paths that start with either `.` or `..` via `is_path_relative_to_cwd()`.